assembl/static2/js/app/components/form/fieldArrayWithActions.jsx

```diff
@@ -2,81 +2,222 @@
 import * as React from 'react';
 import { Button, OverlayTrigger } from 'react-bootstrap';
 import { Translate } from 'react-redux-i18n';
-import { FieldArray } from 'react-final-form-arrays';
+import { FieldArray, type FieldArrayRenderProps } from 'react-final-form-arrays';
+import classNames from 'classnames';
+import range from 'lodash/range';
 
 import { upTooltip, downTooltip } from '../common/tooltips';
 import { createRandomId } from '../../utils/globalFunctions';
+import { MAX_TREE_FORM_LEVEL } from '../../constants';
 
 type Props = {
-  name: string,
   renderFields: Function,
   titleMsgId?: string, // eslint-disable-line react/require-default-props
   tooltips: {
-    addTooltip: React.Node,
-    deleteTooltip: React.Node
+    addTooltip: (props: any) => React.Node,
+    deleteTooltip: (props: any) => React.Node,
+    deleteDisabled?: (props: any) => React.Node
   },
-  withSeparators: boolean
+  withSeparators: boolean,
+  subFieldName?: string,
+  isTree: boolean,
+  level: number,
+  maxLevel: number,
+  minItems: number,
+  parents: Array<number>,
+  parentId: string
 };
 
-const FieldArrayWithActions = ({
-  name,
-  renderFields,
-  titleMsgId,
-  tooltips: { addTooltip, deleteTooltip },
-  withSeparators
-}: Props) => (
-  <FieldArray name={name}>
-    {({ fields }) => (
-      <div>
-        {fields.map((fieldname, idx) => (
-          <div className="form-container" key={fieldname}>
-            {titleMsgId && (
-              <div className="title left">
-                <Translate value={titleMsgId} index={idx + 1} />
-              </div>
-            )}
+type FieldsProps = {
+  fields: FieldArrayRenderProps,
+  onAdd?: (id: string, parentId: string, index: number) => void,
```

Comment:
Why did you add handlers for these actions? Why not hooks for these operations? Eg. `onBeforeAdd`, or `onAfterAdd`, etc. This way, the core of the action itself is controlled by your code, which is safer.